pkg/runtime/pubsub/outbox.go

```diff
@@ -0,0 +1,294 @@
+	"github.com/dapr/kit/logger"
+	"github.com/dapr/kit/retry"
+
+	"github.com/google/uuid"
+)
+
+const (
+	outboxPublishPubsubKey  = "outboxPublishPubsub"
+	outboxPublishTopicKey   = "outboxPublishTopic"
+	outboxPubsubKey         = "outboxPubsub"
+	outboxStateScanDelayKey = "outboxStateScanDelay"
+	outboxStatePrefix       = "outbox"
+	defaultStateScanDelay   = "5s"
+)
+
+var (
+	outboxLogger = logger.NewLogger("dapr.outbox")
+)
+
+type outboxConfig struct {
+	publishPubSub  string
+	publishTopic   string
+	outboxPubsub   string
+	stateScanDelay string
+}
+
+type outboxImpl struct {
+	cloudEventExtractorFn func(map[string]any, string) string
+	getPubsubFn           func(string) (contribPubsub.PubSub, bool)
+	getStateFn            func(string) (state.Store, bool)
+	publishFn             func(context.Context, *contribPubsub.PublishRequest) error
+	outboxStores          map[string]outboxConfig
+	lock                  sync.RWMutex
+}
+
+// NewOutbox returns an instance of an Outbox.
+func NewOutbox(publishFn func(context.Context, *contribPubsub.PublishRequest) error, getPubsubFn func(string) (contribPubsub.PubSub, bool), getStateFn func(string) (state.Store, bool), cloudEventExtractorFn func(map[string]any, string) string) outbox.Outbox {
+	return &outboxImpl{
+		cloudEventExtractorFn: cloudEventExtractorFn,
+		getPubsubFn:           getPubsubFn,
+		getStateFn:            getStateFn,
+		publishFn:             publishFn,
+		lock:                  sync.RWMutex{},
+		outboxStores:          map[string]outboxConfig{},
+	}
+}
+
+// AddOrUpdateOutbox examines a statestore for outbox properties and saves it for later usage in outbox operations.
+func (o *outboxImpl) AddOrUpdateOutbox(stateStore v1alpha1.Component) {
+	var publishPubSub, publishTopicKey, outboxPubsub, stateScanDelay string
+
+	for _, v := range stateStore.Spec.Metadata {
+		switch v.Name {
+		case outboxPublishPubsubKey:
+			publishPubSub = v.Value.String()
+		case outboxPublishTopicKey:
+			publishTopicKey = v.Value.String()
+		case outboxPubsubKey:
+			outboxPubsub = v.Value.String()
+		case outboxStateScanDelayKey:
+			stateScanDelay = v.Value.String()
+		}
+	}
+
+	if publishPubSub != "" && publishTopicKey != "" {
+		o.lock.Lock()
+		defer o.lock.Unlock()
+
+		if outboxPubsub == "" {
+			outboxPubsub = publishPubSub
+		}
+
+		if stateScanDelay == "" {
+			stateScanDelay = defaultStateScanDelay
+		}
+
+		o.outboxStores[stateStore.Name] = outboxConfig{
+			publishPubSub:  publishPubSub,
+			publishTopic:   publishTopicKey,
+			outboxPubsub:   outboxPubsub,
+			stateScanDelay: stateScanDelay,
+		}
+	}
+}
+
+// Enabled returns a bool to indicate if a state store has outbox configured
+func (o *outboxImpl) Enabled(stateStore string) bool {
+	o.lock.RLock()
+	defer o.lock.RUnlock()
+
+	_, ok := o.outboxStores[stateStore]
+	return ok
+}
+
+func transaction() state.TransactionalStateOperation {
+	return state.SetRequest{
+		Key:   outboxStatePrefix + "-" + uuid.New().String(),
+		Value: "0",
```

Comment:
Technically a value isn't needed here at all, but there isn't a deterministic way in Dapr yet to distinguish between key not found and empty value.